### ISSUE TITLE
Use the header bar to display the player controls

### ISSUE THREAD

**Issue by ptitjes** [quodlibet/quodlibet#2297]
Toyed with this idea for 5 minutes... I mean something like this:

![headerbar](https://cloud.githubusercontent.com/assets/64956/23578679/c9c5929a-00dc-11e7-9d08-84a24eeacbbf.png)

Handling full screen could be done ala. Totem, showing the header bar when the pointer is moved.

What do you think ?

**Comment by pschwede** [comment]
Can you implement it as some kind of an optional config?

**Comment by ptitjes** [comment]
@pschwede Yeah, my quick and dirty implementation currently uses `util.is_linux()` to not add the current player controls and the stuff in the header bar, but this can be changed to something more adequate.
(But is there really something outside Gnome ?? ;))

**Comment by pschwede** [comment]
Great!
(I know, I know.)

**Comment by lazka** [comment]
Nice to see the headerbar in action.

Some open issues/questions:

* With the menu bar it looks weird (and this change is mostly about looks/trend..). I guess we'd need to move the menu bar into a menu button somehow.
* By default this has no seek bar widget.
* Headerbars (or client site decorations) kinda work under Windows/macOS, but aren't perfect.

`Gtk.Settings.get_default().props.gtk_dialogs_use_header` should give you True in case the "environment" uses/supports header bars

**Comment by ptitjes** [comment]
> Nice to see the headerbar in action.

Yep, I like it too!


> With the menu bar it looks weird (and this change is mostly about looks/trend..). I guess we'd need to move the menu bar into a menu button somehow.

I was thinking that too while using it. However, I wanted to keep that PR small.


> By default this has no seek bar widget.

Yeah, the actual small seekbar widget is ugly in there. I'd rather put a smaller custom made one if I had to put it there. (But I'd rather deprecate the old one and have the long non-waveform one by default. The short old one is error prone as it keeps the focus.)

> Headerbars (or client site decorations) kinda work under Windows/macOS, but aren't perfect.

I did not know that. Can someone take snapshots of what it looks like ?

**Comment by lazka** [comment]
> I did not know that. Can someone take snapshots of what it looks like ?

They look the same, but gtk+ has to re-implement all the features provided by the "server" site decorations. Window resizing, snapping, shadow drawing, full screen mode etc. And in some cases there is no way to get the same behavior (I've implemented window zooming on macOS, and it's hacky..). Not that it would be a deal breaker, but it needs to be evaluated first.

**Comment by ptitjes** [comment]
On 14/03/2017 13:03, Christoph Reiter wrote:
>
> Not that it would be a deal breaker, but it needs to be evaluated first.
>

And then this design should be made optional, right ?

**Comment by ptitjes** [comment]
@lazka So what do you think ?

- Do I make that an option, instead of testing linux+gnome, for people to test and report ?
- Do you want the seek widget in the headerbar too ?

**Comment by lazka** [comment]
I'd like to see something first which moves the main menu to some menu button. Maybe as a plugin or integrated with your toggle menu plugin.

**Comment by ptitjes** [comment]
Do you mean moving the whole menubar in one only button like what is done in Firefox and Thunderbird ? Or do you mean re-thinking the menu organization like what as been done in Gnome apps ?
I personally would prefer​ the later, but it has to be well thought. I really don't like how this was done in Firefox/Thunderbird and that it is bad UX (hopefully, they kept the menubar toggle in Firefox in addition to the menu button).

I'm affraid that people will shout if we impose them such deep UI changes. But we don't have the main layout extension points to be able to experiment through plugins so that people can try it without being forced to use it if they don't like it... Also we would be able to make it gradually and not in one PR.

**Comment by charlesjt** [comment]
Wow, it looks great. Commonly I am a user of XFCE, but I'm not bothered by the beautiful interface of the headerbar. On the other hand it helps a lot to maintain the coherence of the theme when you want to use a dark variant (The titlebar will always keep the default theme color).

> Yeah, the actual small seekbar widget is ugly in there. I'd rather put a smaller custom made one if I had to put it there. (But I'd rather deprecate the old one and have the long non-waveform one by default. The short old one is error prone as it keeps the focus.)

I very much agree with this. As a user I initially considered it very unintuitive to find the seekbar and always when I recommend Quod Libet I request that they activate the extension of the alternative seekbar for greater convenience.

> I really don't like how this was done in Firefox/Thunderbird and that it is bad UX 

Another example that shows that it is a bad idea is [Pragha](https://github.com/pragha-music-player/pragha). When activating the headerbar, the hamburger menu is filled with too many options. GNOME MPV that also allows switching between a classic interface and one using headerbar does a better job (using popovers + gmenu).

Anyway, the extensions dialog should be integrated as one more tab of the general options and even the navigate menu by integrating to the IMO status bar.

@ptitjes On your last comment, I think that users are already accustomed to applications that present an option to change the interface in a classic way to one using the new GNOME UI, being almost a de facto standard. Ubuntu again used GNOME, so a large number of users would appreciate this kind of integration and there are other desktop environments that prefer the same style as Budgie desktop, Deepin, etc.

**Comment by corebots** [comment]
what is the status of this please? can we now use the player with headerbar controls within GNOME ?

**Comment by CreamyCookie** [comment]
> But can you implement it as some kind of an optional config? Some users don't use Gnome.

Yeah, I agree. Please make this optional. 

One of KDE developers has a lot of good arguments against CSD: [here](https://pointieststick.com/2018/12/18/on-headerbars/) and [followup](https://pointieststick.com/2019/03/06/more-on-headerbars-rebuttals/)

**Comment by declension** [comment]
How popular is this idea now (2021) - are people still keen?

**Comment by bassstorm** [comment]
Functional capabilities are priority over bells & whistles. An opinion from a long time foobar2000 user :)

**Comment by Meriipu** [comment]
I strongly dislike gnome so I would prefer anything like this being configurable (as suggested earlier).
This includes moving the menubar into some button.

**Comment by germe-deb** [comment]
> > But can you implement it as some kind of an optional config? Some users don't use Gnome.
> 
> Yeah, I agree. Please keep this optional.
> 
> One of KDE's developers wrote two great articles discussing a lot of good arguments against CSDs:
> 
> 1. [**On Headerbars**](https://pointieststick.com/2018/12/18/on-headerbars/) - some excerpts:
> 
> * Not universally implementable even by all apps on the same platform
> * Unimplementable in a cross-platform manner
> * Reduced flexibility for users and developers
> * Reduced drag space to move the window
> * App name/window title/filename are omitted or else take up excessive space
> 
> 2. a [followup](https://pointieststick.com/2019/03/06/more-on-headerbars-rebuttals/), where he responds to the feedback he got

"Not universally implementable even by all apps on the same platform"
Quodlibet is an independent app. so, what platform?

"Unimplementable in a cross-platform manner"
I consider this incorrect, as windows now use CSD, and macOS too.
Xorg has support of it, and wayland too

"Reduced flexibility for users and developers"
What flexibility?

"Reduced drag space to move the window"
In gnome this does'nt happens, as the buttons act as the header too, so you can drag the window from a button.

"App name/window title/filename are omitted or else take up excessive space"
This is just an esthetic point, and you don't need the app showing the title all the time.
And this can be an option:
```
Titlebar shows: 
    [x] Song info
    [ ] App name
```

**Comment by Meriipu** [comment]
> > 2. a [followup](https://pointieststick.com/2019/03/06/more-on-headerbars-rebuttals/), where he responds to the feedback he got
> 
> "Not universally implementable even by all apps on the same platform" Quodlibet is an independent app. so, what platform?
> 

If it is an independent app, why work hard to have it have "a consistent style with gnome" ?  c:

Quite a few people would not consider it being configurable as simply "better" as this might imply it being less-than-necessary.

**Comment by germe-deb** [comment]
> > > 3\. a [followup](https://pointieststick.com/2019/03/06/more-on-headerbars-rebuttals/), where he responds to the feedback he got
> > 
> > 
> > "Not universally implementable even by all apps on the same platform" Quodlibet is an independent app. so, what platform?
> 
> If it is an independent app, why work hard to have it have "a consistent style with gnome" ? c:
> 
> Quite a few people would not consider CSD being configurable as simply "better", as this might imply configurability being "nice to have but less-than-necessary".

well, quodlibet needs a face wash now. It looks like a 2010 app basically, and the CSD is a small step to modernization

It will make quodlibet more atractive to new people, and it will bring a better experience to a certain percentage of users using it already

**Comment by slosd** [comment]
I'm taking a stab at this (see #4122), properly incorporating items of the menu bar. Looking for general feedback and maybe some ideas for what to do with the vertical row of buttons on the left (see PR for details).
![headerbar2](https://user-images.githubusercontent.com/99086/186963107-62670a29-be4a-4c13-b725-c56520c11415.png)

**Comment by haevalencia** [comment]
Nice, it looks like a good interface refresh. I love Quod Libet, but the app feels out of place, even on XFCE where most apps have migrated from menu bars to header bars.
